File: sdks/ts/memory/src/connector/slack.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Slack connector: fetches channel messages, reconstructs threads,
 * downloads file attachments, and converts Slack mrkdwn to standard
 * markdown. Implements the Connector interface from the connector
 * framework (P5-1).
 *
 * Rate limiting: Slack Tier 3 APIs (~50 req/min) are respected via the
 * shared RateLimiter. 429 responses trigger backoff with Retry-After.
 */

import type { Connector, ConnectorDocument, SyncCursor } from './types.js'
import { RateLimiter } from './types.js'
import {
  convertMrkdwn,
  formatDate,
  parseSlackTimestamp,
  readResponseWithLimit,
  validateDownloadURL,
} from './slack_helpers.js'

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

export type SlackConnectorConfig = {
  readonly botToken: string
  readonly channels: readonly string[]
  readonly includeThreads?: boolean
  readonly includeFiles?: boolean
  readonly maxFileSize?: number
  readonly oldestTimestamp?: string
  readonly pollIntervalMs?: number
  /** Injectable fetch function for testing. */
  readonly fetchFn?: typeof fetch
}

// ---------------------------------------------------------------------------
// Slack API response types
// ---------------------------------------------------------------------------

type SlackResponse = {
  readonly ok: boolean
  readonly error?: string
  readonly messages?: readonly SlackMessage[]
  readonly response_metadata?: { readonly next_cursor?: string }
}

type SlackMessage = {
  readonly type: string
  readonly user: string
  readonly text: string
  readonly ts: string
  readonly thread_ts?: string
  readonly reply_count?: number
  readonly files?: readonly SlackFile[]
}

type SlackFile = {
  readonly id: string
  readonly name: string
  readonly mimetype: string
  readonly size: number
  readonly url_private_download: string
}

type SlackUserResponse = {
  readonly ok: boolean
  readonly user?: {
    readonly real_name?: string
    readonly name?: string
  }
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_MAX_FILE_SIZE = 50 * 1024 * 1024 // 50 MB
const DEFAULT_POLL_INTERVAL_MS = 5 * 60 * 1000 // 5 minutes
const API_TIMEOUT_MS = 30_000
const FILE_DOWNLOAD_TIMEOUT_MS = 60_000
const MAX_RESPONSE_BYTES = 10 * 1024 * 1024 // 10 MB
const MAX_API_RETRIES = 5

// ---------------------------------------------------------------------------
// Slack connector
// ---------------------------------------------------------------------------

export class SlackConnector implements Connector {
  readonly name = 'slack' as const

  private config: SlackConnectorConfig = {
    botToken: '',
    channels: [],
    includeThreads: true,
    includeFiles: true,
    maxFileSize: DEFAULT_MAX_FILE_SIZE,
    pollIntervalMs: DEFAULT_POLL_INTERVAL_MS,
  }

  private readonly rateLimiter = new RateLimiter({
    maxTokens: 50,
    refillRate: 50 / 60, // ~0.833 tokens/sec for Tier 3
  })

  private readonly userCache = new Map<string, string>()
  private stopController: AbortController | undefined
  private readonly fetchFn: typeof fetch

  constructor(config?: Partial<SlackConnectorConfig>) {
    this.fetchFn = config?.fetchFn ?? globalThis.fetch.bind(globalThis)
    if (config) {
      this.config = {
        ...this.config,
        ...config,
        maxFileSize: config.maxFileSize ?? DEFAULT_MAX_FILE_SIZE,
        pollIntervalMs: config.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS,
      }
    }
  }

  async configure(config: Record<string, unknown>): Promise<void> {
    const botToken = typeof config['botToken'] === 'string' ? config['botToken'] : ''
    if (!botToken) {
      throw new Error('slack: botToken is required')
    }

    const channelsRaw = typeof config['channels'] === 'string' ? config['channels'] : ''
    if (!channelsRaw) {
      throw new Error('slack: at least one channel is required')
    }

    const channels = channelsRaw.split(',').map(ch => ch.trim()).filter(Boolean)
    if (channels.length === 0) {
      throw new Error('slack: at least one channel is required')
    }

    let resolvedMaxFileSize = this.config.maxFileSize
    const maxFileSizeRaw = config['maxFileSize']
    if (typeof maxFileSizeRaw === 'string') {
      const parsed = Number.parseInt(maxFileSizeRaw, 10)
      if (Number.isNaN(parsed)) {
        throw new Error('slack: invalid maxFileSize')
      }
      resolvedMaxFileSize = parsed
    } else if (typeof maxFileSizeRaw === 'number') {
      resolvedMaxFileSize = maxFileSizeRaw
    }

    const resolvedOldestTimestamp =
      typeof config['oldestTimestamp'] === 'string'
        ? config['oldestTimestamp']
        : this.config.oldestTimestamp

    const includeThreadsRaw = config['includeThreads']
    const includeThreads =
      typeof includeThreadsRaw === 'boolean'
        ? includeThreadsRaw
        : (typeof includeThreadsRaw === 'string' ? includeThreadsRaw !== 'false' : (this.config.includeThreads ?? true))

    const includeFilesRaw = config['includeFiles']
    const includeFiles =
      typeof includeFilesRaw === 'boolean'
        ? includeFilesRaw
        : (typeof includeFilesRaw === 'string' ? includeFilesRaw !== 'false' : (this.config.includeFiles ?? true))

    this.config = {
      ...this.config,
      botToken,
      channels,
      includeThreads,
      includeFiles,
      ...(resolvedMaxFileSize !== undefined ? { maxFileSize: resolvedMaxFileSize } : {}),
      ...(resolvedOldestTimestamp !== undefined ? { oldestTimestamp: resolvedOldestTimestamp } : {}),
    }
  }

  async *fetchAll(signal: AbortSignal): AsyncIterable<ConnectorDocument> {
    yield* this.fetchMessages(signal, '')
  }

  async *fetchSince(signal: AbortSignal, cursor: SyncCursor): AsyncIterable<ConnectorDocument> {
    yield* this.fetchMessages(signal, cursor.value)
  }

  async start(signal: AbortSignal): Promise<void> {
    this.stopController = new AbortController()
    const combinedSignal = AbortSignal.any([signal, this.stopController.signal])
    let lastCursor = this.config.oldestTimestamp ?? ''

    while (!combinedSignal.aborted) {
      let latestTS = ''

      for await (const doc of this.fetchMessages(combinedSignal, lastCursor)) {
        const ts = doc.metadata['ts']
        if (ts && ts > latestTS) {
          latestTS = ts
        }
        // In the P5-1 framework, start() blocks and the framework
        // handles document dispatch internally. Documents fetched
        // during the continuous loop are consumed here.
        void doc
      }

      if (latestTS) {
        lastCursor = latestTS
      }

      await this.sleep(this.config.pollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS, combinedSignal)
    }
  }

  async stop(): Promise<void> {
    this.stopController?.abort()
  }

  // -------------------------------------------------------------------------
  // Internal: message fetching
  // -------------------------------------------------------------------------

  private async *fetchMessages(signal: AbortSignal, oldest: string): AsyncIterable<ConnectorDocument> {
    for (const channelId of this.config.channels) {
      yield* this.fetchChannelMessages(signal, channelId, oldest)
    }
  }

  private async *fetchChannelMessages(
    signal: AbortSignal,
    channelId: string,
    oldest: string,
  ): AsyncIterable<ConnectorDocument> {
    let cursor = ''

    while (true) {
      await this.rateLimiter.acquire(1, signal)

      const resp = await this.callConversationsHistory(signal, channelId, oldest, cursor)

      for (const msg of resp.messages ?? []) {
        yield this.messageToDocument(channelId, msg)

        // Fetch thread replies.
        if (this.config.includeThreads && (msg.reply_count ?? 0) > 0 && msg.thread_ts) {
          const threadDoc = await this.fetchThread(signal, channelId, msg)
          if (threadDoc) {
            yield threadDoc
          }
        }

        // Fetch file attachments.
        if (this.config.includeFiles && msg.files && msg.files.length > 0) {
          for (const file of msg.files) {
            if (file.size > (this.config.maxFileSize ?? DEFAULT_MAX_FILE_SIZE)) {
              continue
            }
            const fileDoc = await this.downloadFile(signal, channelId, file)
            if (fileDoc) {
              yield fileDoc
            }
          }
        }
      }

      const nextCursor = resp.response_metadata?.next_cursor
      if (!nextCursor) {
        break
      }
      cursor = nextCursor
    }
  }

  // -------------------------------------------------------------------------
  // Internal: thread fetching
  // -------------------------------------------------------------------------

  private async fetchThread(
    signal: AbortSignal,
    channelId: string,
    parent: SlackMessage,
  ): Promise<ConnectorDocument | undefined> {
    const allReplies: SlackMessage[] = []
    let cursor = ''

    while (true) {
      await this.rateLimiter.acquire(1, signal)

      const resp = await this.callConversationsReplies(signal, channelId, parent.thread_ts ?? parent.ts, cursor)

      if (resp.messages) {
        allReplies.push(...resp.messages)
      }

      const nextCursor = resp.response_metadata?.next_cursor
      if (!nextCursor) {
        break
      }
      cursor = nextCursor
    }

    const content = this.buildThreadDocument(parent, allReplies)
    const ts = parseSlackTimestamp(parent.ts)

    return {
      externalId: `${channelId}:thread:${parent.thread_ts ?? parent.ts}`,
      content,
      mime: 'text/markdown',
      title: `Thread in ${channelId}`,
      metadata: {
        channel: channelId,
        user: parent.user,
        ts: parent.ts,
        thread_ts: parent.thread_ts ?? parent.ts,
        reply_count: String(parent.reply_count ?? 0),
        source: 'slack',
        type: 'thread',
      },
      modifiedAt: ts,
    }
  }

  private buildThreadDocument(parent: SlackMessage, replies: readonly SlackMessage[]): string {
    const lines: string[] = []

    lines.push(`## Thread: ${convertMrkdwn(parent.text)}`)
    lines.push('')

    // Include the parent message as the first entry.
    const parentUser = this.resolveUserName(parent.user)
    const parentTS = parseSlackTimestamp(parent.ts)
    lines.push(`**${parentUser}** (${formatDate(parentTS)}):`)
    lines.push(convertMrkdwn(parent.text))
    lines.push('')

    for (const reply of replies) {
      // Skip the parent message (Slack includes it in replies).
      if (reply.ts === parent.ts) {
        continue
      }
      const userName = this.resolveUserName(reply.user)
      const ts = parseSlackTimestamp(reply.ts)
      const dateStr = formatDate(ts)
      lines.push(`**${userName}** (${dateStr}):`)
      lines.push(convertMrkdwn(reply.text))
      lines.push('')
    }

    return lines.join('\n').trimEnd()
  }

  // -------------------------------------------------------------------------
  // Internal: file download
  // -------------------------------------------------------------------------

  private async downloadFile(
    signal: AbortSignal,
    channelId: string,
    file: SlackFile,
  ): Promise<ConnectorDocument | undefined> {
    // Validate the download URL against SSRF before fetching.
    validateDownloadURL(file.url_private_download)

    await this.rateLimiter.acquire(1, signal)

    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), FILE_DOWNLOAD_TIMEOUT_MS)
    const combinedSignal = AbortSignal.any([signal, controller.signal])

    try {
      const response = await this.fetchFn(file.url_private_download, {
        headers: { Authorization: `Bearer ${this.config.botToken}` },
        signal: combinedSignal,
      })

      if (!response.ok) {
        return undefined
      }

      const maxSize = this.config.maxFileSize ?? DEFAULT_MAX_FILE_SIZE
      const content = await readResponseWithLimit(response, maxSize)

      if (content.length > maxSize) {
        return undefined
      }

      return {
        externalId: `${channelId}:file:${file.id}`,
        content,
        mime: file.mimetype,
        title: file.name,
        metadata: {
          channel: channelId,
          file_id: file.id,
          filename: file.name,
          filetype: file.mimetype,
          file_size: String(file.size),
          source: 'slack',
          type: 'file',
        },
        modifiedAt: new Date(),
      }
    } finally {
      clearTimeout(timeout)
    }
  }

  // -------------------------------------------------------------------------
  // Internal: Slack API calls
  // -------------------------------------------------------------------------

  private async callConversationsHistory(
    signal: AbortSignal,
    channelId: string,
    oldest: string,
    cursor: string,
  ): Promise<SlackResponse> {
    const params = new URLSearchParams({ channel: channelId, limit: '200' })
    if (oldest) params.set('oldest', oldest)
    if (cursor) params.set('cursor', cursor)

    return this.callSlackAPI(signal, 'https://slack.com/api/conversations.history', params)
  }

  private async callConversationsReplies(
    signal: AbortSignal,
    channelId: string,
    threadTs: string,
    cursor: string,
  ): Promise<SlackResponse> {
    const params = new URLSearchParams({ channel: channelId, ts: threadTs, limit: '200' })
    if (cursor) params.set('cursor', cursor)

    return this.callSlackAPI(signal, 'https://slack.com/api/conversations.replies', params)
  }

  private async callSlackAPI(
    signal: AbortSignal,
    endpoint: string,
    params: URLSearchParams,
  ): Promise<SlackResponse> {
    const requestUrl = `${endpoint}?${params.toString()}`
    let backoffMs = 0

    for (let attempt = 0; attempt < MAX_API_RETRIES; attempt++) {
      if (backoffMs > 0) {
        await this.sleep(backoffMs, signal)
      }

      const controller = new AbortController()
      const timeout = setTimeout(() => controller.abort(), API_TIMEOUT_MS)
      const combinedSignal = AbortSignal.any([signal, controller.signal])

      try {
        const response = await this.fetchFn(requestUrl, {
          headers: {
            Authorization: `Bearer ${this.config.botToken}`,
            'Content-Type': 'application/x-www-form-urlencoded',
          },
          signal: combinedSignal,
        })

        // Handle rate limiting (HTTP 429) with exponential backoff.
        if (response.status === 429) {
          const retryAfter = response.headers.get('Retry-After')
          const waitSecs = retryAfter ? Number.parseInt(retryAfter, 10) : 5
          const baseMs = (Number.isNaN(waitSecs) ? 5 : Math.max(1, waitSecs)) * 1000

          if (attempt === MAX_API_RETRIES - 1) {
            throw new Error(`slack API rate limited after ${MAX_API_RETRIES} retries`)
          }

          // Exponential backoff: use Retry-After as base, double on each subsequent retry.
          backoffMs = baseMs * (1 << attempt)
          continue
        }

        if (!response.ok) {
          throw new Error(`slack API returned HTTP ${response.status}`)
        }

        // Read response with size limit to prevent OOM.
        const bodyBuffer = await readResponseWithLimit(response, MAX_RESPONSE_BYTES)
        const data = JSON.parse(bodyBuffer.toString('utf-8')) as SlackResponse

        if (!data.ok) {
          throw new Error(`slack API error: ${data.error ?? 'unknown'}`)
        }

        return data
      } finally {
        clearTimeout(timeout)
      }
    }

    throw new Error(`slack API: exhausted all ${MAX_API_RETRIES} retry attempts`)
  }

  // -------------------------------------------------------------------------
  // Internal: user resolution
  // -------------------------------------------------------------------------

  private resolveUserName(userId: string): string {
    if (!userId) return 'unknown'

    const cached = this.userCache.get(userId)
    if (cached) return cached

    // Return the user ID for now -- async resolution would complicate
    // the synchronous buildThreadDocument flow. The user cache is
    // populated lazily via fetchUserName calls where feasible.
    return userId
  }

  // -------------------------------------------------------------------------
  // Internal: message -> document conversion
  // -------------------------------------------------------------------------

  private messageToDocument(channelId: string, msg: SlackMessage): ConnectorDocument {
    const ts = parseSlackTimestamp(msg.ts)
    const content = convertMrkdwn(msg.text)

    const metadata: Record<string, string> = {
      channel: channelId,
      user: msg.user,
      ts: msg.ts,
      source: 'slack',
      type: 'message',
    }
    if (msg.thread_ts) {
      metadata['thread_ts'] = msg.thread_ts
    }
    if (msg.reply_count && msg.reply_count > 0) {
      metadata['reply_count'] = String(msg.reply_count)
    }

    return {
      externalId: `${channelId}:${msg.ts}`,
      content,
      mime: 'text/markdown',
      title: `Message in ${channelId}`,
      metadata,
      modifiedAt: ts,
    }
  }

  // -------------------------------------------------------------------------
  // Helpers
  // -------------------------------------------------------------------------

  private async sleep(ms: number, signal: AbortSignal): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      if (signal.aborted) {
        reject(signal.reason ?? new Error('aborted'))
        return
      }
      const timer = setTimeout(resolve, ms)
      const onAbort = (): void => {
        clearTimeout(timer)
        reject(signal.reason ?? new Error('aborted'))
      }
      signal.addEventListener('abort', onAbort, { once: true })
    })
  }
}

// Re-export helpers from slack_helpers for consumers that import from slack.ts.
export { convertMrkdwn, parseSlackTimestamp } from './slack_helpers.js'

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

/**
 * Factory function for creating a Slack connector with the given
 * configuration.
 */
export function createSlackConnector(config?: Partial<SlackConnectorConfig>): SlackConnector {
  return new SlackConnector(config)
}

File: go/connector/slack.go
// SPDX-License-Identifier: Apache-2.0

package connector

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"
)

// maxAPIRetries is the maximum number of retry attempts for rate-limited
// (HTTP 429) API responses before returning an error.
const maxAPIRetries = 5

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

// SlackConnectorConfig holds the configuration for the Slack connector.
type SlackConnectorConfig struct {
	BotToken        string
	Channels        []string
	IncludeThreads  bool
	IncludeFiles    bool
	MaxFileSize     int64
	OldestTimestamp string
	PollInterval    time.Duration
	Logger          *slog.Logger
	HTTPClient      HTTPDoer
}

// defaultSlackConfig returns a SlackConnectorConfig with sensible defaults
// applied for unset fields.
func defaultSlackConfig() SlackConnectorConfig {
	return SlackConnectorConfig{
		IncludeThreads: true,
		IncludeFiles:   true,
		MaxFileSize:    50 * 1024 * 1024, // 50 MB
		PollInterval:   5 * time.Minute,
	}
}

// ---------------------------------------------------------------------------
// HTTP abstraction (for testability)
// ---------------------------------------------------------------------------

// HTTPDoer is the minimal interface for issuing HTTP requests. Both
// *http.Client and test doubles satisfy it.
type HTTPDoer interface {
	Do(req *http.Request) (*http.Response, error)
}

// ---------------------------------------------------------------------------
// Slack API response types
// ---------------------------------------------------------------------------

type slackResponse struct {
	OK               bool           `json:"ok"`
	Error            string         `json:"error"`
	Messages         []slackMessage `json:"messages"`
	ResponseMetadata slackRespMeta  `json:"response_metadata"`
}

type slackRespMeta struct {
	NextCursor string `json:"next_cursor"`
}

type slackMessage struct {
	Type       string      `json:"type"`
	User       string      `json:"user"`
	Text       string      `json:"text"`
	TS         string      `json:"ts"`
	ThreadTS   string      `json:"thread_ts"`
	ReplyCount int         `json:"reply_count"`
	Files      []slackFile `json:"files"`
}

type slackFile struct {
	ID                 string `json:"id"`
	Name               string `json:"name"`
	MIMEType           string `json:"mimetype"`
	Size               int64  `json:"size"`
	URLPrivateDownload string `json:"url_private_download"`
}

type slackUserResponse struct {
	OK   bool      `json:"ok"`
	User slackUser `json:"user"`
}

type slackUser struct {
	RealName string `json:"real_name"`
	Name     string `json:"name"`
}

// ---------------------------------------------------------------------------
// Slack connector
// ---------------------------------------------------------------------------

// SlackConnector implements the Connector interface for Slack workspace
// integration. It fetches channel messages, reconstructs threads,
// downloads file attachments, and converts Slack mrkdwn to standard
// markdown.
type SlackConnector struct {
	config       SlackConnectorConfig
	rateLimiter  *RateLimiter
	userCache    map[string]string
	userMu       sync.RWMutex
	stopCh       chan struct{}
	stopped      bool
	mu           sync.Mutex
	logger       *slog.Logger
	httpClient   HTTPDoer
	urlValidator func(string) error // SSRF validation; defaults to validateDownloadURL
}

// NewSlackConnector creates a new Slack connector with the given config.
// Missing config fields are filled with defaults.
func NewSlackConnector(cfg SlackConnectorConfig) *SlackConnector {
	defaults := defaultSlackConfig()

	if cfg.MaxFileSize <= 0 {
		cfg.MaxFileSize = defaults.MaxFileSize
	}
	if cfg.PollInterval <= 0 {
		cfg.PollInterval = defaults.PollInterval
	}
	if cfg.Logger == nil {
		cfg.Logger = slog.Default()
	}
	if cfg.HTTPClient == nil {
		cfg.HTTPClient = &http.Client{Timeout: 30 * time.Second}
	}

	return &SlackConnector{
		config: cfg,
		// Slack Tier 3: ~50 requests per minute = ~0.833 per second
		rateLimiter:  NewRateLimiter(50, 50.0/60.0),
		userCache:    make(map[string]string),
		stopCh:       make(chan struct{}),
		logger:       cfg.Logger,
		httpClient:   cfg.HTTPClient,
		urlValidator: validateDownloadURL,
	}
}

// Name returns "slack".
func (c *SlackConnector) Name() string { return "slack" }

// Configure validates and applies connector-specific configuration.
// Accepts map[string]any to match the P5-1 Connector interface.
func (c *SlackConnector) Configure(config map[string]any) error {
	token, _ := config["botToken"].(string)
	if token == "" {
		return fmt.Errorf("slack: botToken is required")
	}
	c.config.BotToken = token

	channelsRaw, _ := config["channels"].(string)
	if channelsRaw == "" {
		return fmt.Errorf("slack: at least one channel is required")
	}
	c.config.Channels = strings.Split(channelsRaw, ",")
	for i, ch := range c.config.Channels {
		c.config.Channels[i] = strings.TrimSpace(ch)
	}

	if v, ok := config["includeThreads"].(bool); ok {
		c.config.IncludeThreads = v
	} else if v, ok := config["includeThreads"].(string); ok {
		c.config.IncludeThreads = v != "false"
	}

	if v, ok := config["includeFiles"].(bool); ok {
		c.config.IncludeFiles = v
	} else if v, ok := config["includeFiles"].(string); ok {
		c.config.IncludeFiles = v != "false"
	}

	if v, ok := config["maxFileSize"].(string); ok {
		size, err := strconv.ParseInt(v, 10, 64)
		if err != nil {
			return fmt.Errorf("slack: invalid maxFileSize: %w", err)
		}
		c.config.MaxFileSize = size
	} else if v, ok := config["maxFileSize"].(float64); ok {
		c.config.MaxFileSize = int64(v)
	} else if v, ok := config["maxFileSize"].(int); ok {
		c.config.MaxFileSize = int64(v)
	}

	if v, ok := config["oldestTimestamp"].(string); ok {
		c.config.OldestTimestamp = v
	}

	return nil
}

// FetchAll performs a full sync of all configured channels.
func (c *SlackConnector) FetchAll(ctx context.Context) (<-chan ConnectorDocument, <-chan error) {
	return c.fetchMessages(ctx, "")
}

// FetchSince performs an incremental sync starting from the cursor value
// (a Slack message timestamp).
func (c *SlackConnector) FetchSince(ctx context.Context, cursor SyncCursor) (<-chan ConnectorDocument, <-chan error) {
	return c.fetchMessages(ctx, cursor.Value)
}

// Start begins a continuous sync loop that polls at the configured
// interval. Blocks until the context is cancelled or Stop is called.
// Returns the first fatal error encountered, or nil on clean shutdown.
func (c *SlackConnector) Start(ctx context.Context) error {
	var lastCursor string
	if c.config.OldestTimestamp != "" {
		lastCursor = c.config.OldestTimestamp
	}

	ticker := time.NewTicker(c.config.PollInterval)
	defer ticker.Stop()

	// Run once immediately, then on each tick.
	for {
		docs, errs := c.fetchMessages(ctx, lastCursor)
		var latestTS string

		for doc := range docs {
			ts, tsOK := doc.Metadata["ts"]
			if tsOK && ts > latestTS {
				latestTS = ts
			}
		}

		if err, ok := <-errs; ok && err != nil {
			return fmt.Errorf("slack: sync error: %w", err)
		}

		if latestTS != "" {
			lastCursor = latestTS
		}

		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-c.stopCh:
			return nil
		case <-ticker.C:
		}
	}
}

// Stop gracefully stops the continuous sync loop.
func (c *SlackConnector) Stop() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if !c.stopped {
		c.stopped = true
		close(c.stopCh)
	}
	return nil
}

// ---------------------------------------------------------------------------
// Internal: message fetching
// ---------------------------------------------------------------------------

func (c *SlackConnector) fetchMessages(ctx context.Context, oldest string) (<-chan ConnectorDocument, <-chan error) {
	docCh := make(chan ConnectorDocument, 100)
	errCh := make(chan error, 1)

	go func() {
		defer close(docCh)
		defer close(errCh)

		for _, channelID := range c.config.Channels {
			if err := c.fetchChannelMessages(ctx, channelID, oldest, docCh); err != nil {
				select {
				case errCh <- fmt.Errorf("slack: channel %s: %w", channelID, err):
				default:
				}
				return
			}
		}
	}()

	return docCh, errCh
}

func (c *SlackConnector) fetchChannelMessages(
	ctx context.Context,
	channelID string,
	oldest string,
	docCh chan<- ConnectorDocument,
) error {
	cursor := ""
	for {
		if err := c.rateLimiter.Acquire(ctx, 1); err != nil {
			return err
		}

		resp, err := c.callConversationsHistory(ctx, channelID, oldest, cursor)
		if err != nil {
			return fmt.Errorf("conversations.history: %w", err)
		}

		for _, msg := range resp.Messages {
			doc := c.messageToDocument(channelID, msg)
			select {
			case docCh <- doc:
			case <-ctx.Done():
				return ctx.Err()
			}

			// Fetch thread replies if enabled and the message is a thread parent.
			if c.config.IncludeThreads && msg.ReplyCount > 0 && msg.ThreadTS != "" {
				threadDoc, err := c.fetchThread(ctx, channelID, msg)
				if err != nil {
					c.logger.Warn("slack: failed to fetch thread",
						"channel", channelID,
						"thread_ts", msg.ThreadTS,
						"error", err,
					)
					continue
				}
				select {
				case docCh <- threadDoc:
				case <-ctx.Done():
					return ctx.Err()
				}
			}

			// Fetch file attachments if enabled.
			if c.config.IncludeFiles && len(msg.Files) > 0 {
				for _, file := range msg.Files {
					if file.Size > c.config.MaxFileSize {
						c.logger.Warn("slack: skipping file exceeding size limit",
							"file_id", file.ID,
							"file_name", file.Name,
							"size", file.Size,
							"max_size", c.config.MaxFileSize,
						)
						continue
					}
					fileDoc, err := c.downloadFile(ctx, channelID, file)
					if err != nil {
						c.logger.Warn("slack: failed to download file",
							"file_id", file.ID,
							"error", err,
						)
						continue
					}
					select {
					case docCh <- fileDoc:
					case <-ctx.Done():
						return ctx.Err()
					}
				}
			}
		}

		if resp.ResponseMetadata.NextCursor == "" {
			break
		}
		cursor = resp.ResponseMetadata.NextCursor
	}
	return nil
}

// ---------------------------------------------------------------------------
// Internal: thread fetching
// ---------------------------------------------------------------------------

func (c *SlackConnector) fetchThread(
	ctx context.Context,
	channelID string,
	parent slackMessage,
) (ConnectorDocument, error) {
	var allReplies []slackMessage
	cursor := ""

	for {
		if err := c.rateLimiter.Acquire(ctx, 1); err != nil {
			return ConnectorDocument{}, err
		}

		resp, err := c.callConversationsReplies(ctx, channelID, parent.ThreadTS, cursor)
		if err != nil {
			return ConnectorDocument{}, fmt.Errorf("conversations.replies: %w", err)
		}

		allReplies = append(allReplies, resp.Messages...)

		if resp.ResponseMetadata.NextCursor == "" {
			break
		}
		cursor = resp.ResponseMetadata.NextCursor
	}

	content := c.buildThreadDocument(parent, allReplies)
	ts := parseSlackTimestamp(parent.TS)

	return ConnectorDocument{
		ExternalID: fmt.Sprintf("%s:thread:%s", channelID, parent.ThreadTS),
		Content:    []byte(content),
		MIME:       "text/markdown",
		Title:      fmt.Sprintf("Thread in %s", channelID),
		Metadata: map[string]string{
			"channel":     channelID,
			"user":        parent.User,
			"ts":          parent.TS,
			"thread_ts":   parent.ThreadTS,
			"reply_count": strconv.Itoa(parent.ReplyCount),
			"source":      "slack",
			"type":        "thread",
		},
		ModifiedAt: ts,
	}, nil
}

func (c *SlackConnector) buildThreadDocument(parent slackMessage, replies []slackMessage) string {
	var sb strings.Builder

	parentUser := c.resolveUserName(parent.User)
	parentTS := parseSlackTimestamp(parent.TS)
	sb.WriteString("## Thread: ")
	sb.WriteString(ConvertMrkdwn(parent.Text))
	sb.WriteString("\n\n")

	// Include the parent message as the first entry.
	sb.WriteString(fmt.Sprintf("**%s** (%s):\n", parentUser, formatTimestamp(parentTS)))
	sb.WriteString(ConvertMrkdwn(parent.Text))
	sb.WriteString("\n\n")

	for _, reply := range replies {
		// Skip the parent message itself (Slack includes it in replies).
		if reply.TS == parent.TS {
			continue
		}
		replyUser := c.resolveUserName(reply.User)
		ts := parseSlackTimestamp(reply.TS)
		sb.WriteString(fmt.Sprintf("**%s** (%s):\n", replyUser, formatTimestamp(ts)))
		sb.WriteString(ConvertMrkdwn(reply.Text))
		sb.WriteString("\n\n")
	}

	return strings.TrimRight(sb.String(), "\n")
}

// ---------------------------------------------------------------------------
// Internal: file download
// ---------------------------------------------------------------------------

func (c *SlackConnector) downloadFile(
	ctx context.Context,
	channelID string,
	file slackFile,
) (ConnectorDocument, error) {
	// Validate the download URL against SSRF before fetching.
	if err := c.urlValidator(file.URLPrivateDownload); err != nil {
		return ConnectorDocument{}, err
	}

	if err := c.rateLimiter.Acquire(ctx, 1); err != nil {
		return ConnectorDocument{}, err
	}

	reqCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, file.URLPrivateDownload, nil)
	if err != nil {
		return ConnectorDocument{}, fmt.Errorf("create file request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.config.BotToken)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return ConnectorDocument{}, fmt.Errorf("download file: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return ConnectorDocument{}, fmt.Errorf("download file: HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(io.LimitReader(resp.Body, c.config.MaxFileSize+1))
	if err != nil {
		return ConnectorDocument{}, fmt.Errorf("read file body: %w", err)
	}
	if int64(len(body)) > c.config.MaxFileSize {
		return ConnectorDocument{}, fmt.Errorf("file %s exceeds maximum size of %d bytes", file.Name, c.config.MaxFileSize)
	}

	return ConnectorDocument{
		ExternalID: fmt.Sprintf("%s:file:%s", channelID, file.ID),
		Content:    body,
		MIME:       file.MIMEType,
		Title:      file.Name,
		Metadata: map[string]string{
			"channel":   channelID,
			"file_id":   file.ID,
			"filename":  file.Name,
			"filetype":  file.MIMEType,
			"file_size": strconv.FormatInt(file.Size, 10),
			"source":    "slack",
			"type":      "file",
		},
		ModifiedAt: time.Now(),
	}, nil
}

// ---------------------------------------------------------------------------
// Internal: Slack API calls
// ---------------------------------------------------------------------------

func (c *SlackConnector) callConversationsHistory(
	ctx context.Context,
	channelID string,
	oldest string,
	cursor string,
) (slackResponse, error) {
	params := url.Values{
		"channel": {channelID},
		"limit":   {"200"},
	}
	if oldest != "" {
		params.Set("oldest", oldest)
	}
	if cursor != "" {
		params.Set("cursor", cursor)
	}

	return c.callSlackAPI(ctx, "https://slack.com/api/conversations.history", params)
}

func (c *SlackConnector) callConversationsReplies(
	ctx context.Context,
	channelID string,
	threadTS string,
	cursor string,
) (slackResponse, error) {
	params := url.Values{
		"channel": {channelID},
		"ts":      {threadTS},
		"limit":   {"200"},
	}
	if cursor != "" {
		params.Set("cursor", cursor)
	}

	return c.callSlackAPI(ctx, "https://slack.com/api/conversations.replies", params)
}

func (c *SlackConnector) callSlackAPI(
	ctx context.Context,
	endpoint string,
	params url.Values,
) (slackResponse, error) {
	reqURL := endpoint + "?" + params.Encode()
	backoff := time.Duration(0)

	for attempt := range maxAPIRetries {
		if backoff > 0 {
			select {
			case <-ctx.Done():
				return slackResponse{}, ctx.Err()
			case <-time.After(backoff):
			}
		}

		reqCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
		req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, reqURL, nil)
		if err != nil {
			cancel()
			return slackResponse{}, fmt.Errorf("create request: %w", err)
		}
		req.Header.Set("Authorization", "Bearer "+c.config.BotToken)
		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

		resp, err := c.httpClient.Do(req)
		if err != nil {
			cancel()
			return slackResponse{}, fmt.Errorf("http request: %w", err)
		}

		// Handle rate limiting (HTTP 429) with exponential backoff.
		if resp.StatusCode == http.StatusTooManyRequests {
			retryAfter := resp.Header.Get("Retry-After")
			waitSecs, parseErr := strconv.Atoi(retryAfter)
			if parseErr != nil || waitSecs <= 0 {
				waitSecs = 5
			}
			resp.Body.Close()
			cancel()

			if attempt == maxAPIRetries-1 {
				return slackResponse{}, fmt.Errorf("slack API rate limited after %d retries", maxAPIRetries)
			}

			// Exponential backoff: use Retry-After as base, double on each subsequent retry.
			backoff = time.Duration(waitSecs) * time.Second * (1 << attempt)
			continue
		}

		if resp.StatusCode != http.StatusOK {
			resp.Body.Close()
			cancel()
			return slackResponse{}, fmt.Errorf("slack API returned HTTP %d", resp.StatusCode)
		}

		body, readErr := io.ReadAll(io.LimitReader(resp.Body, 10*1024*1024)) // 10 MB limit
		resp.Body.Close()
		cancel()
		if readErr != nil {
			return slackResponse{}, fmt.Errorf("read response: %w", readErr)
		}

		var result slackResponse
		if err := json.Unmarshal(body, &result); err != nil {
			return slackResponse{}, fmt.Errorf("decode response: %w", err)
		}

		if !result.OK {
			return slackResponse{}, fmt.Errorf("slack API error: %s", result.Error)
		}

		return result, nil
	}

	return slackResponse{}, fmt.Errorf("slack API: exhausted all %d retry attempts", maxAPIRetries)
}
